### ISSUE TITLE
Support for GitLab

### ISSUE THREAD

**Issue by NatoBoram** [anuraghazra/github-readme-stats#1563]
![stale](https://avatars.githubusercontent.com/in/1724?s=88&v=4)

* https://github.com/anuraghazra/github-readme-stats/issues/203

> I would be nice when you can also query data also from https://gitlab.com (as option in your API). I've already filed a issue on GitLab.com for [profile README.md for GitLab](https://gitlab.com/gitlab-org/gitlab/-/issues/232157).
> 
> Here is a link to the [GraphQL API of GitLab](https://docs.gitlab.com/ee/api/graphql)

* https://github.com/anuraghazra/github-readme-stats/issues/753

> I was using the readme status for my account and I consider how it can combine results from GitHub and GitLab. At this time it seems this readme-stats is working just from GitHub results.

**Comment by rickstaa** [comment]
@NatoBoram, I like your feature request. However, I am not sure if it falls within the project scope that @anuraghazra envisioned. We are currently working on more pressing feature requests and bugs and are not implementing this shortly. However, if @anuraghazra thinks this is within the project scope, somebody can create a pull request for it. I will leave this issue open to check the support of the community.

**Comment by hamishmb** [comment]
Would you accept a pull request for this? I might do this one day when I have some free time.

**Comment by hamishmb** [comment]
Okay, that seems fair, I shall wait for the overhaul to finish.

**Comment by rickstaa** [comment]
@hamishmb, on the second tought, the changes in #1633 are related to the SVG and, therefore, would not conflict with your proposed PR too much since they are related to the fetchers. As a result, I think you can already implement this feature to use it on your fork. 

Whether the PR gets merged into the main branch depends on how much it complicates the codebase and @anuraghazra opinion. I am happy to review an implementation proposal to see if it complicates the codebase before you start to implement it. 

@anuraghazra, do you think supporting other Git providers like GitLab is in the project's scope, or would you prefer we stay at GitHub?

**Comment by hamishmb** [comment]
Okay. It's probably going to be quite a while before I have the time/headspace anyway, so I'll probably still end up doing it afterwards.

**Comment by rickstaa** [comment]
@hamishmb No hurries! I fully understand. We're all maintaining this project in our free time 💪. I just wanted to check with @anuraghazra before you implemented something that we will not merge.

**Comment by hamishmb** [comment]
Oh yeah no worries all good, still good to have the information either way :)

**Comment by anuraghazra** [comment]
@NatoBoram thanks for the feature request. 

But I don't think we will be supporting GitLab. 
Reason being, I want to keep the stats to GitHub only and there will be other overheads in adding a new platform: 

- We need to add access tokens for gitlab
- Another 2 or 3 cards to maintain
- We already have a shit ton of load in the vercel instance / serverless function

**Comment by hamishmb** [comment]
Ah okay, that's totally fair enough. Thanks for making the enquiry for me though.

**Comment by rickstaa** [comment]
@anuraghazra Makes sense! Maybe we can release the core (i.e. [common](https://github.com/anuraghazra/github-readme-stats/tree/master/src/common)) components as a `core` package under https://github.com/readme-stats? People can submit their cards like `gitlab-readme-stats` and `codestats-readme-stats` to be maintained under the organization. After the summer holidays, I am happy to work on `github-readme-stats/gitlab-readme-stats`.

**Comment by rickstaa** [comment]
Having this organizational structure greatly benefits the developability and open-source nature of the project since we have better permission management giving us the ability to have a larger team. We could also put the `themes` and `translations` in separate repositories uncluttering the pull requests. 🤔

**Comment by hamishmb** [comment]
Still doesn't solve the issue of where to run it, if your existing VM instance (or whatever it is) doesn't have any extra capacity.

**Comment by rickstaa** [comment]
@hamishmb correct it only splits the capacity over multiple Vercel instances, repositories and PAT.

**Comment by hamishmb** [comment]
Out of interest, what is PAT?

**Comment by rickstaa** [comment]
O sorry for using the abbreviation without explanation. It refers to [Github Personal Access  Token (PAT)](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token). We use it in the public Vercel instance to interact with the GraphQL API (see https://github.com/anuraghazra/github-readme-stats/issues/1772 and https://github.com/anuraghazra/github-readme-stats/blob/master/src/common/retryer.js).

**Comment by rickstaa** [comment]
@hamishmb I just created a PR to export the modules and functions from the `common` folder of GRS as an npm package. If #2151 is merged, you can easily create your own GitLab-based readme stats card using the `github-readme-stats` package.

**Comment by hamishmb** [comment]
Excellent, thanks for telling me.

I've done a little web development, so would need to learn more, but this might make a fun project someday :+1:

**Comment by Zo-Bro-23** [comment]
I know this is a little old, but if #2409 gets merged, you can easily add GitLab stats, and perhaps create a PR too (maybe I will try doing it myself later on)!

**Comment by rickstaa** [comment]
> I know this is a little old, but if #2409 gets merged, you can easily add GitLab stats, and perhaps create a PR too (maybe I will try doing it myself later on)!

Amazing, GitLab support would be great! 👍

**Comment by lapaz17** [comment]
Is there any progress on this?

**Comment by Zo-Bro-23** [comment]
It will be possible to implement after #2473 gets merged.
